samza-kafka/src/main/scala/org/apache/samza/checkpoint/kafka/KafkaCheckpointManager.scala

```diff
@@ -156,53 +158,76 @@ class KafkaCheckpointManager(checkpointSpec: KafkaStreamSpec,
     }
 
     val envelope = new OutgoingMessageEnvelope(checkpointSsp, keyBytes, msgBytes)
-    val retryBackoff: ExponentialSleepStrategy = new ExponentialSleepStrategy
 
-    val startTime = System.currentTimeMillis()
-    retryBackoff.run(
-      loop => {
-        systemProducer.send(taskName.getTaskName, envelope)
-        systemProducer.flush(taskName.getTaskName) // make sure it is written
+    // Used for exponential backoff retries on failure in sending messages through producer.
+    val startTimeInMillis: Long = System.currentTimeMillis()
+    var sleepTimeInMillis: Long = 1000
+    var producerException: Exception = null
+    while ((System.currentTimeMillis() - startTimeInMillis) <= MaxRetryDurationInMillis) {
+      val currentProducer = producerRef.get()
+      try {
+        currentProducer.send(taskName.getTaskName, envelope)
+        currentProducer.flush(taskName.getTaskName) // make sure it is written
         debug(s"Wrote checkpoint: $checkpoint for task: $taskName")
-        loop.done
-      },
-
-      (exception, loop) => {
-        if ((System.currentTimeMillis() - startTime) >= MaxRetryDurationMs) {
-          error(s"Exhausted $MaxRetryDurationMs milliseconds when writing checkpoint: $checkpoint for task: $taskName.")
-          throw new SamzaException(s"Exception when writing checkpoint: $checkpoint for task: $taskName.", exception)
-        } else {
+        return
+      } catch {
+        case exception: Exception => {
+          producerException = exception
           warn(s"Retrying failed checkpoint write to key: $key, checkpoint: $checkpoint for task: $taskName", exception)
+          // TODO: Remove this producer recreation logic after SAMZA-1393.
+          val newProducer: SystemProducer = getSystemProducer()
+          producerCreationLock.synchronized {
+            if (producerRef.compareAndSet(currentProducer, newProducer)) {
+              info(s"Stopping the checkpoint SystemProducer")
+              currentProducer.stop()
+              info(s"Recreating the checkpoint SystemProducer")
+              // SystemProducer contract is that clients call register(taskName) followed by start
+              // before invoking writeCheckpoint, readCheckpoint API. Hence list of taskName are not
+              // expected to change during the producer recreation.
+              for (taskName <- taskNames) {
+                debug(s"Registering the taskName: $taskName with SystemProducer")
+                newProducer.register(taskName.getTaskName)
+              }
+              newProducer.start()
+            } else {
+              info("Producer instance was recreated by other thread. Retrying with it.")
+              newProducer.stop()
+            }
+          }
         }
       }
-    )
+      sleepTimeInMillis = sleepTimeInMillis * 2
```

Comment:
Is there specific benefit of exponential backoff instead of constant backoff? It seems that the probability of the broker/network issue being fixed is equally distributed over time and thus it is reasonable to use constant backoff.

If we want to use exponential backoff is better, it seems good to keep the existing logic of `ExponentialSleepStrategy` which upper bounds the maximum delay at 10 sec. Otherwise the code may sleep for minutes which does not seem reasonable.

Reply:
Regarding `Exponential backoff vs constant backoff`, if the goal is to prevent thundering herd of requests to Kafka service, then it does not seem to be enforced here because the underlying KafkaProducer already retries request with constant backoff. So it is not clear why we are using exponential backoff. I am OK to skip this issue since you prefer to keep it.

Regarding the second issue, I am not suggesting that we should re-use the ExponentialSleepStrategy class. I am just saying that is is reasonable to re-use the existing logic in ExponentialSleepStrategy which upper bounds the retry backoff to 10 sec. Thanks for making the change.